# "Introduction to Random Strings" solution optimization using NumPy

## Data set:

In [1]:
dna_seq = "GAATTATGGAAATAGTCATCAAGCTTGAGTCTGGACTCAACACTTAGGAAAATCCTATAGGAGCTGGGGTTCAAACATATCGCAA"
cg_content = "0.084 0.136 0.164 0.214 0.291 0.339 0.354 0.442 0.460 0.533 0.567 0.636 0.699 0.721 0.758 0.807 0.853 0.929"
huge_cg_content = " ".join([cg_content] * 1000)

## Solution using vanilla python:

In [2]:
from math import log10

def vanilla_solution(dna_seq, cg_content_string):
    cg_content_list = [float(item) for item in cg_content_string.split(' ')]
    result = []

    cg_count = dna_seq.count('G') + dna_seq.count('C')
    at_count = len(dna_seq) - cg_count
    
    for cg_prob in cg_content_list:
        prob = (cg_count * log10(cg_prob / 2)) + (at_count * log10((1 - cg_prob) / 2))
        result.append(prob)
    return result

- Time complexity: $O(N)$
- Space complexity: $O(N)$

In [3]:
%timeit vanilla_solution(dna_seq, huge_cg_content)

33.2 ms ± 8.14 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [4]:
%timeit vanilla_solution(dna_seq, cg_content)

26.8 μs ± 1.69 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [5]:
answer = vanilla_solution(dna_seq, cg_content)
print(answer)

[-64.10538474827405, -58.28502588042745, -56.25033864842683, -53.686931489556116, -51.432230011581154, -50.730485458765955, -50.59958030481232, -50.564848327752756, -50.701701603293486, -51.74366361566322, -52.504842915057935, -54.65383713329048, -57.46843438079922, -58.691936264024335, -61.10421297494659, -65.19042669149155, -70.40209849958563, -85.26083957098176]


## Solution using numpy:

In [6]:
import numpy as np
from math import log10

def numpy_solution(dna_seq, cg_content_string):
    gen = (float(item) for item in cg_content_string.split(' '))
    cg_content_arr = np.fromiter(gen, dtype=float)

    gc_count = dna_seq.count('G') + dna_seq.count('C')
    at_count = len(dna_seq) - gc_count

    cg_probs = np.log10(cg_content_arr / 2)
    at_probs = np.log10((1 - cg_content_arr) / 2)
    
    result = (gc_count * cg_probs) + (at_count * at_probs)

    return result.tolist()

- Time complexity: $O(N)$
- Space complexity: $O(N)$

In [7]:
%timeit numpy_solution(dna_seq, huge_cg_content)

14.6 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [8]:
%timeit numpy_solution(dna_seq, cg_content)

52.7 μs ± 4.78 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [9]:
answer = numpy_solution(dna_seq, cg_content)
print(answer)

[-64.10538474827405, -58.28502588042745, -56.25033864842683, -53.686931489556116, -51.432230011581154, -50.730485458765955, -50.59958030481232, -50.564848327752756, -50.701701603293486, -51.74366361566322, -52.504842915057935, -54.65383713329048, -57.46843438079922, -58.691936264024335, -61.10421297494659, -65.19042669149155, -70.40209849958563, -85.26083957098176]


**Summary:** Vanilla Python outperforms NumPy when processing small inputs (e.g., the original ```cg_content``` string). In this case, Python took approximately **26.8 μs** compared to NumPy's **52.7 μs**. This is because the overhead of initializing NumPy arrays, converting data types, and calling C-extensions outweighs the actual computation time for a few elements. NumPy demonstrates its true power on the "huge" dataset (scaled by 1000x). The NumPy implementation completed the task in **14.6 ms**, while the vanilla solution took **33.2 ms** making NumPy more than twice as fast.